### ISSUE TITLE
Add "Dhizuku" method in the installation method list

### ISSUE THREAD

**Issue by XieyilinMain** [ImranR98/Obtainium#1611]
**Describe the feature**

https://github.com/iamr0s/Dhizuku

Add "Dhizuku" method in the installation method list to make Obtainium compatible with Dhizuku when installing APK.

**Additional context**

Dhizuku is an open-source DeviceOwner provider which is more stable than Shizuku on non-rooted devices. It does **NOT** need any authorization after a reboot, and is less likely to be killed in the background situation.

Some apps like [Amarok](https://f-droid.org/packages/deltazero.amarok.foss/) and InstallerX are already supported this method.

[Dhizuku API sample](https://github.com/iamr0s/Dhizuku-API)

**Comment by h0cheung** [comment]
InstallerX supports silent installation. So a simple way is adding "legacy" installation method, then we can use InstallerX in Obtainium, to automatically install applications.

**Comment by picatsv** [comment]
as this feature request has been stale for a while i wanted to see if i could find an alternative way to make the two work, and turns out there is a stopgap solution: patching obtainium with lspatch and the xposed dhizuku compatibility module lets me successfully update and install apps through dhizuku! may be a step in the right direction for an implementation?

**Comment by ManiaciaChao** [comment]
I recently bought a China-stock phone and discovered that the ROM has modified the framework to intercept normal `pm install` requests[^1]. As a result, silent installation via _Shizuku_ or _Sui_ no longer works—this is especially frustrating since I have a large number of APKs tracked in _Obtainium_.

That said, installing APKs under **Device Owner** permissions still works flawlessly. In this case, integrating with [Dhizuku](https://github.com/iamr0s/Dhizuku) has proven to be an effective and fast workaround.

I'm willing to take a look into this and hopefully implement a demo by the weekend.

[^1]: For a detailed write-up on reverse engineering the stock package installer, see [this post](https://leadroyal.cn/p/1151/).

**Comment by re7gog** [comment]
Ok, I will do it

**Comment by ManiaciaChao** [comment]
> Ok, I will do it

Here's a clever workaround that avoids directly including Dhizuku in the code repository.

When sending the Intent, specify the class name and activity to invoke an external installer. The authorization process (from Root/Shizuku/Dhizuku) can be entirely handled by a third-party installer (e.g., InstallerX).

You can refer to the pull request I just submitted to Droid-ify as an example: https://github.com/Droid-ify/client/pull/995

**Comment by re7gog** [comment]
> > Ok, I will do it
> 
> Here's a clever workaround that avoids directly including Dhizuku in the code repository.
> 
> When sending the Intent, specify the class name and activity to invoke an external installer. The authorization process (from Root/Shizuku/Dhizuku) can be entirely handled by a third-party installer (e.g., InstallerX).
> 
> You can refer to the pull request I just submitted to Droid-ify as an example: [Droid-ify/client#995](https://github.com/Droid-ify/client/pull/995)

Окак

**Comment by re7gog** [comment]
![Image](https://github.com/user-attachments/assets/e822329c-ab35-48db-aacf-401d217c23ca)
Dhizuku is funny

**Comment by h0cheung** [comment]
> > Ok, I will do it
> 
> Here's a clever workaround that avoids directly including Dhizuku in the code repository.
> 
> When sending the Intent, specify the class name and activity to invoke an external installer. The authorization process (from Root/Shizuku/Dhizuku) can be entirely handled by a third-party installer (e.g., InstallerX).
> 
> You can refer to the pull request I just submitted to Droid-ify as an example: https://github.com/Droid-ify/client/pull/995

In most devices, just adding the legacy installation method is enough. Only in some Chinese ROM this is needed.

**Comment by ManiaciaChao** [comment]
> > > Ok, I will do it
> > 
> > 
> > Here's a clever workaround that avoids directly including Dhizuku in the code repository.
> > When sending the Intent, specify the class name and activity to invoke an external installer. The authorization process (from Root/Shizuku/Dhizuku) can be entirely handled by a third-party installer (e.g., InstallerX).
> > You can refer to the pull request I just submitted to Droid-ify as an example: [Droid-ify/client#995](https://github.com/Droid-ify/client/pull/995)
> 
> In most devices, just adding the legacy installation method is enough. Only in some Chinese ROM this is needed.

Yes, I totally agree.

Incorporating the Dhizuku API into the codebase isn't a wise decision—especially considering that its maintainer has announced he will no longer be actively involved with the project ([source](https://t.me/s/InstallerX)). Simply sending a `PACKAGE_INSTALL` intent should suffice.

That said, some Chinese ROMs intercept this intent and invoke their own custom installers instead. I don't have access to Obtainium's user distribution statistics (and I don't believe such data is currently collected), so I can't accurately estimate how many users are on CN ROMs.

Specifying an installer component directly or using `Intent.createChooser(intent)` would be a minimal code change, but it has the potential to work around a number of related issues.

**Comment by ManiaciaChao** [comment]
> ![Image](https://github.com/user-attachments/assets/e822329c-ab35-48db-aacf-401d217c23ca) Dhizuku is funny

Xaxa, I think this behavior is triggered by the Device Owner feature (I experienced the same thing when using OwnDroid). It seems like a default policy issue.

Maybe it's worth considering @h0cheung’s suggestion to just send the intent directly? It's waaaaaay easier to implement and avoids the complications introduced by Dhizuku.

**Comment by re7gog** [comment]
Oh yes, I checked myself, in auto mode it can

**Comment by ManiaciaChao** [comment]
> > > ![Image](https://github.com/user-attachments/assets/e822329c-ab35-48db-aacf-401d217c23ca) Dhizuku is funny
> > 
> > 
> > Xaxa, I think this behavior is triggered by the Device Owner feature (I experienced the same thing when using OwnDroid). It seems like a default policy issue.
> > Maybe it's worth considering [@h0cheung](https://github.com/h0cheung)’s suggestion to just send the intent directly? It's waaaaaay easier to implement and avoids the complications introduced by Dhizuku.
> 
> Question is, can installerx do it's job in background without asking user, on problematic ROMs? If yes, it's a good idea. I made this module mainly for Chinese phones when used one some time ago

Yes, InstallerX allows users to set up separate configurations for different apps. 

For instance, I’ve configured an "Auto notification" mode for Droid-ify—meaning it will automatically install the APK when a notification is triggered. Droid-ify sends the intent, and InstallerX installs the APK silently (though it still shows a notification).

I’ve verified this behavior on both Xiaomi (not the latest version, as my test device is a bit old) and Oppo devices.

**Comment by re7gog** [comment]
Ok, it seems like it's better to write a new flutter module for using selectable installers and replace existing one

**Comment by ManiaciaChao** [comment]
> Ok, it seems like it's better to write a new flutter module for using selectable installers and replace existing one

Could keep the current implementation. Rooted users don’t need such a InstallerX workaround anyway—though it's worth noting that most recent phones available on the CN market are nearly un-rootable.

Implementing a Flutter module is such a good idea. It would allow other projects to reuse the installation logic more easily.

**Comment by re7gog** [comment]
Using module is necessary, because otherwise, installation process gets killed in the background.
Maybe it's better to do both things: complete dhizuku implementation, because it's half done, and add an option to just use 3rd party installers

**Comment by picatsv** [comment]
> Incorporating the Dhizuku API into the codebase isn't a wise decision—especially considering that its maintainer has announced he will no longer be actively involved with the project ([source](https://t.me/s/InstallerX)).

do note that the repos are currently maintained still by the current maintainer and still see commits regularly, this is basically just a leadership change

**Comment by re7gog** [comment]
Done! Just update package, it's fully backwards compatible and automatically prefers dhizuku over shizuku

**Comment by re7gog** [comment]
Note: it's probably better to remove Spoof as Play market function since it's not very useful and unsupported when using Dhizuku

**Comment by ManiaciaChao** [comment]
> Note: it's probably better to remove Spoof as Play market function since it's not very useful and unsupported when using Dhizuku

Good to know! I'll try it this weekend.

**Comment by re7gog** [comment]
@ImranR98, can you please update plugin dependency

**Comment by DwainZwerg** [comment]
> Note: it's probably better to remove Spoof as Play market function since it's not very useful and unsupported when using Dhizuku

@re7gog I don't think it makes sense, because it can be quite helpful (e.g. for banking apps). It would make sense to either a) detect whether Dhizuku is being used (and then not display the option) or b) make it more clear that this only works with Shizuku.
